### DISCUSSION TITLE
How do we run with a group license in a cloud CI environment?

### DISCUSSION THREAD

**Discussion by ryanpennellblacksage** [dashingsoft/pyarmor#1473] [General]
Hi, 

I was reading the documentation for obfuscating scripts in a CI pipeline.  We are using CircleCI for launching our CI docker images. 

Since the docker host can be anything, how do we obfuscate without binding our group license to a specific ephemeral cloud machine?

When running https://pyarmor.readthedocs.io/en/stable/how-to/register.html?highlight=CI#registering-in-docker-or-ci-pipeline, we get an error that it is improper for group license

**Comment by ryanpennellblacksage** [comment]
As a side note, if we can't run unlimited CI builds, then we would likely want to return the pyarmor license.  We need to be able to run more than 100 CI builds per day

**Comment by jondy** [comment]
Do you check this link
https://pyarmor.readthedocs.io/en/latest/licenses.html#check-device-for-group-license

If you purchase group licnese in this month, an notice email about group license should be sent separately.

**Comment by jondy** [comment]
If pyarmor group license doesn't work for your case, please do not activate it and ask for refund. If it's used, refund will be rejected.

**Comment by ryanpennellblacksage** [comment]
Hi @jondy, 

We're pretty confused.  When purchasing, we were under the impression the group license worked in both an online (unlimited) and offline context (100 device limit), and parts of the [pyarmor documentation](https://pyarmor.readthedocs.io/en/latest/how-to/register.html?highlight=ci#registering-in-docker-or-ci-pipeline) make it seem like this is the case.  Our use case requires both because oftentimes, we build in areas without internet connectivity.

When running with a cloud CI provider where we don't have host configuration access, would a pro license work?  For reference, our cloud CI provider is CircleCI.  Is there any way to get pyarmor group running on CircleCI?  If not and a pro license works, could we convince you to provide a pro license free of charge? 

Unfortunately, our group license has already been activated

**Comment by jondy** [comment]
This section [pyarmor documentation](https://pyarmor.readthedocs.io/en/latest/how-to/register.html?highlight=ci#registering-in-docker-or-ci-pipeline) is under section `using Pyarmor pro/basic license`.

I think you misunderstand group license, I also add more notes in last 30 days, search `Hints for 3 licenses` in [license typs](https://pyarmor.readthedocs.io/en/latest/licenses.html)

I have no ideas about CircleCI, if it could be configured to keep machine id got by pyarmor same, it works. Otherwise, the possible solution is running docker containers in the physical device or cloud server (like ECS in Aliyun), obfuscated scripts in those containers and take them as input of CI.
